File: services/control-plane/internal/transport/http/handler/invites.go
// Package handler — invite surface (Phase 3).
//
// Three protected endpoints under /v1/orgs/{id}/invites: issue, list, revoke.
// Two public endpoints under /v1/invites/{token}: get (landing page) and claim
// (sets session cookie). RBAC: issue + revoke are owner-only; list is admin or
// owner. The token in the URL is the raw 32-byte base64url value emailed to
// the invitee; it is hashed before any DB lookup.
package handler

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/nexis-eco/nexis/services/control-plane/internal/domain"
	"github.com/nexis-eco/nexis/services/control-plane/internal/platform/config"
	appmw "github.com/nexis-eco/nexis/services/control-plane/internal/transport/http/middleware"
)

// issueInviteReq is the body of POST /v1/orgs/{id}/invites.
type issueInviteReq struct {
	Email string `json:"email"`
	Role  string `json:"role"`
}

// InviteIssue mints an invite for the caller's org. The {id} URL param must
// match the caller's org_id — we don't trust cross-org issuance even from an
// owner. Returns 201 with token_prefix (first 8 chars of the raw token) — the
// full token only lands in the email so the prefix is safe to show in toasts.
func InviteIssue(auth domain.AuthProvider, audit domain.AuditWriter) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		if chi.URLParam(r, "id") != princ.OrgID {
			httpJSON(w, http.StatusForbidden, map[string]string{"error": "forbidden"})
			return
		}
		var req issueInviteReq
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			httpJSON(w, http.StatusBadRequest, map[string]string{"error": "bad json"})
			return
		}
		role := domain.Role(req.Role)
		if role != domain.RoleAdmin && role != domain.RoleMember {
			httpJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid role"})
			return
		}
		token, err := auth.IssueInvite(r.Context(), princ, req.Email, role)
		if err != nil {
			if errors.Is(err, domain.ErrConflict) {
				httpJSON(w, http.StatusConflict, map[string]string{"error": err.Error()})
				return
			}
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		if audit != nil {
			_ = audit.Write(r.Context(), princ, "invite.issued", req.Email, map[string]any{"role": req.Role})
		}
		prefix := token
		if len(prefix) > 8 {
			prefix = prefix[:8]
		}
		httpJSON(w, http.StatusCreated, map[string]string{"token_prefix": prefix})
	}
}

// InviteList wires GET /v1/orgs/{id}/invites. Includes claimed + expired rows
// — the UI badges them; the caller does not need to know about each row's
// lifecycle state to render the table.
func InviteList(auth domain.AuthProvider) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		if chi.URLParam(r, "id") != princ.OrgID {
			httpJSON(w, http.StatusForbidden, map[string]string{"error": "forbidden"})
			return
		}
		rows, err := auth.ListInvites(r.Context(), princ)
		if err != nil {
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		out := make([]map[string]any, 0, len(rows))
		for _, i := range rows {
			out = append(out, map[string]any{
				"token_hash": toHex(i.TokenHash),
				"email":      i.Email,
				"role":       string(i.Role),
				"expires_at": i.ExpiresAt.UTC().Format(time.RFC3339),
				"claimed_at": nullableTimeStr(i.ClaimedAt),
			})
		}
		httpJSON(w, http.StatusOK, out)
	}
}

// InviteRevoke wires DELETE /v1/orgs/{id}/invites/{token_hash}. token_hash is
// the hex string surfaced by the list endpoint.
func InviteRevoke(auth domain.AuthProvider, audit domain.AuditWriter) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		if chi.URLParam(r, "id") != princ.OrgID {
			httpJSON(w, http.StatusForbidden, map[string]string{"error": "forbidden"})
			return
		}
		tokenHashHex := chi.URLParam(r, "token_hash")
		if err := auth.RevokeInvite(r.Context(), princ, tokenHashHex); err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				httpJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
				return
			}
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		if audit != nil {
			_ = audit.Write(r.Context(), princ, "invite.revoked", tokenHashHex, nil)
		}
		w.WriteHeader(http.StatusNoContent)
	}
}

// InviteGet wires GET /v1/invites/{token} (public). Returns the org + role +
// inviter email so the claim landing page can render a coherent prompt.
func InviteGet(auth domain.AuthProvider) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		info, err := auth.GetInviteInfo(r.Context(), chi.URLParam(r, "token"))
		if err != nil {
			httpJSON(w, http.StatusNotFound, map[string]string{"error": err.Error()})
			return
		}
		httpJSON(w, http.StatusOK, map[string]any{
			"org": map[string]string{
				"id":   info.OrgID,
				"name": info.OrgName,
				"slug": info.OrgSlug,
			},
			"role":          string(info.Role),
			"inviter_email": info.InviterEmail,
		})
	}
}

// claimReq is the body of POST /v1/invites/{token}/claim. Password is required
// only for brand-new users; for existing users the field is ignored.
type claimReq struct {
	Password string `json:"password"`
}

// InviteClaim wires POST /v1/invites/{token}/claim (public). On success the
// session cookie is set so the invitee lands on the dashboard already signed in.
func InviteClaim(auth domain.AuthProvider, cfg config.Config) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req claimReq
		_ = json.NewDecoder(r.Body).Decode(&req)
		st, err := auth.ClaimInvite(r.Context(), chi.URLParam(r, "token"), req.Password)
		if err != nil {
			httpJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
			return
		}
		setSessionCookie(w, cfg, st)
		w.WriteHeader(http.StatusOK)
	}
}

File: apps/web/lib/invites.ts
// Phase 3 Stage 8/10 — invites SDK.
//
// Used by:
//   * Settings → Members & Roles (Stage 8): admin-only list/issue/revoke
//     under /v1/orgs/{id}/invites.
//   * Invite claim flow (Stage 10): public lookup + claim under
//     /v1/invites/{token}{,/claim}.
//
// All admin calls forward the user's session cookie via credentials:"include".
// The public claim endpoints accept no auth — the raw token in the URL is
// the proof of intent.

const API =
  typeof window === "undefined"
    ? process.env.API_URL_INTERNAL ?? process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8080"
    : process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8080";

// PendingInvite mirrors the JSON returned by GET /v1/orgs/{id}/invites.
// Fields match the snake_case the Go handler emits via json tags.
export type PendingInvite = {
  token_hash: string;
  email: string;
  role: "admin" | "member";
  expires_at: string;
  claimed_at?: string | null;
};

// InviteInfo matches the public GET /v1/invites/{token} response.
export type InviteInfo = {
  org: { id: string; name: string; slug: string };
  role: "admin" | "member";
  inviter_email: string;
};

async function unwrap(r: Response): Promise<never> {
  const body = await r.json().catch(() => ({}) as Record<string, unknown>);
  const msg =
    typeof body === "object" && body !== null && "error" in body
      ? String((body as { error: unknown }).error)
      : r.statusText;
  throw new Error(msg);
}

export const invites = {
  list: async (orgId: string): Promise<PendingInvite[]> => {
    const r = await fetch(`${API}/v1/orgs/${orgId}/invites`, {
      credentials: "include",
    });
    if (!r.ok) await unwrap(r);
    return r.json();
  },
  issue: async (
    orgId: string,
    email: string,
    role: "admin" | "member",
  ): Promise<{ token_prefix: string }> => {
    const r = await fetch(`${API}/v1/orgs/${orgId}/invites`, {
      method: "POST",
      credentials: "include",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ email, role }),
    });
    if (!r.ok) await unwrap(r);
    return r.json();
  },
  revoke: async (orgId: string, tokenHash: string): Promise<void> => {
    const r = await fetch(`${API}/v1/orgs/${orgId}/invites/${tokenHash}`, {
      method: "DELETE",
      credentials: "include",
    });
    if (!r.ok && r.status !== 204) await unwrap(r);
  },
  // Public — no cookie required. Looks up the org + role + inviter info
  // for the landing page before the invitee picks a password.
  get: async (token: string): Promise<InviteInfo> => {
    const r = await fetch(`${API}/v1/invites/${token}`);
    if (!r.ok) await unwrap(r);
    return r.json();
  },
  // Public — sets the nexis_session cookie on success.
  claim: async (token: string, password: string): Promise<void> => {
    const r = await fetch(`${API}/v1/invites/${token}/claim`, {
      method: "POST",
      credentials: "include",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ password }),
    });
    if (!r.ok) await unwrap(r);
  },
};
